### PR TITLE
Allow updating particular branches

### PR DESCRIPTION
**Pull request by feedmeapples**
## What was changed

- Starts accepting Submodule name and Branch as the inputs (ex. `temporal`)
- Checks out and updates the submodule for a given branch (ex. `v1.1.0`)
- Commits the changes into a branch with the same name

## Why?

Automating patch releases

## Checklist

1. Closes 

2. How was this tested:

Tested the code with personal repos and simulating the changes:
https://github.com/feedmeapples/gha-parent
https://github.com/feedmeapples/gha-child

3. Any docs updates needed?

### CODE REVIEW

**Review by dnr — COMMENTED** [review]
So, this strategy works alright if one submodule has release branches (in this case it would be temporal). On each release branch it'll start from the tctl main branch at the point where that release branch was created, and then never update it along that branch. It'll be a big pain if we need to patch tctl on multiple release branches, but I guess we can hope we never have to do that.

File: .github/workflows/update-submodules.yml [hunk 1]
```diff
@@ -17,21 +24,33 @@ jobs:
...
 
-      - uses: actions/setup-go@v2
-        with:
-          go-version: 1.17
+      - name: Create and switch branch
+        if: github.event.inputs.branch != 'master' # remain on 'main' branch if target branch is 'master'
+        env:
+          BRANCH: ${{ github.event.inputs.branch }}
+        run: |
+          git switch $BRANCH || git switch -c $BRANCH
```
[the remark this answers is not in the document]
Comment (feedmeapples) [reply .github/workflows/update-submodules.yml:36]: GHA stops the execution if there is git error, 
ex. here is a `git switch` failure from one of the early action executions https://github.com/feedmeapples/gha-parent/runs/5178238896?check_suite_focus=true#step:3:12

File: .github/workflows/update-submodules.yml [hunk 2]
```diff
@@ -17,21 +24,33 @@ jobs:
...
+          SUBMODULE: ${{ github.event.inputs.repo }}
+          BRANCH: ${{ github.event.inputs.branch }}
         run: |
           git config --local user.name 'Temporal Data (cicd)'
           git config --local user.email 'commander-data@temporal.io'
           git remote set-url origin https://x-access-token:${{ secrets.COMMANDER_DATA_TOKEN }}@github.com/${{ github.repository }}
           git add .
-          git commit -m "Update submodules" && git push || echo "No changes to commit"
+          git commit -m "Update $SUBMODULE submodule for branch $BRANCH"
+          git push --set-upstream origin $BRANCH || echo "No changes to commit"
```
[the remark this answers is not in the document]
Comment (feedmeapples) [reply .github/workflows/update-submodules.yml:56]: if we create a new branch with `git switch -c $BRANCH` (ex. `release/v1.15.0`), then it's not going to have the upstream target, so i'm setting it explicitly

File: .github/workflows/update-submodules.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (feedmeapples) [reply .github/workflows/update-submodules.yml]: done

File: .github/workflows/update-submodules.yml [hunk 3]
```diff
@@ -17,21 +24,33 @@ jobs:
...
+          git submodule update --init --recursive --remote
 
       - name: Commit update
+        env:
+          SUBMODULE: ${{ github.event.inputs.repo }}
+          BRANCH: ${{ github.event.inputs.branch }}
         run: |
           git config --local user.name 'Temporal Data (cicd)'
           git config --local user.email 'commander-data@temporal.io'
           git remote set-url origin https://x-access-token:${{ secrets.COMMANDER_DATA_TOKEN }}@github.com/${{ github.repository }}
```
[the remark this answers is not in the document]
Comment (feedmeapples) [reply .github/workflows/update-submodules.yml]: done

File: .github/workflows/update-submodules.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (dnr) [reply .github/workflows/update-submodules.yml]: Hmm, but you're pushing to a target explicitly. The man page (and my understanding) is that the state updated by `--set-upstream` is in the local clone, and is used by git pull/push with no specified branch. So I don't think you need it here. Does it work without it, and if not, what behavior/error do you get?

File: .github/workflows/update-submodules.yml [hunk 4]
```diff
@@ -17,21 +24,35 @@ jobs:
...
+        env:
+          CURRENT_REPO: ${{ github.repository }}
+          TOKEN: ${{ secrets.COMMANDER_DATA_TOKEN }}
+          SUBMODULE: ${{ github.event.inputs.repo }}
+          BRANCH: ${{ github.event.inputs.branch }}
         run: |
           git config --local user.name 'Temporal Data (cicd)'
           git config --local user.email 'commander-data@temporal.io'
-          git remote set-url origin https://x-access-token:${{ secrets.COMMANDER_DATA_TOKEN }}@github.com/${{ github.repository }}
+          git remote set-url origin https://x-access-token:$TOKEN@github.com/$CURRENT_REPO
```
Comment (dnr) [inline .github/workflows/update-submodules.yml:55]: put the whole url in quotes

File: .github/workflows/update-submodules.yml [hunk 5]
```diff
@@ -17,21 +24,35 @@ jobs:
...
         run: |
-          git pull --recurse-submodules
-          git submodule update --remote --recursive
+          git config --file=.gitmodules submodule."$SUBMODULE".branch "$BRANCH"
+          git submodule update --init --recursive --remote
 
       - name: Commit update
+        env:
+          CURRENT_REPO: ${{ github.repository }}
+          TOKEN: ${{ secrets.COMMANDER_DATA_TOKEN }}
```
Comment (dnr) [inline .github/workflows/update-submodules.yml:49]: why are there two different tokens? can they be combined to one?

**Review by dnr — COMMENTED** [review]
So this doesn't actually build images for the release branch commits yet, right? Is that going to come in another PR?

File: .github/workflows/update-submodules.yml (hunk 5 above)
Comment (feedmeapples) [reply .github/workflows/update-submodules.yml:49]: short answer is to stick to the minimal permissions necessary for a particular action.
Ideally we would only use `GITHUB_TOKEN` as it is new for every workflow run, however it doesn't allow to write into a protected branch, so the other commander token allows to do that

Comment (feedmeapples) [reply .github/workflows/update-submodules.yml:49]: yea it wasn't. Added the build trigger in the current PR https://github.com/temporalio/docker-builds/pull/23/commits/bc9706130e7761bb844dd93253862f3d7a2de5ee

File: .github/workflows/update-submodules.yml (hunk 4 above)
Comment (feedmeapples) [reply .github/workflows/update-submodules.yml]: done

File: .github/workflows/update-submodules.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (feedmeapples) [reply .github/workflows/update-submodules.yml]: right, it's unnecessary. Removed the flag

File: .github/workflows/update-submodules.yml [hunk 6]
```diff
@@ -17,21 +24,33 @@ jobs:
...
-          persist-credentials: false
+          persist-credentials: true
+          token: ${{ secrets.GITHUB_TOKEN }}
+          fetch-depth: 0
 
-      - uses: actions/setup-go@v2
-        with:
-          go-version: 1.17
+      - name: Create and switch branch
+        if: github.event.inputs.branch != 'master' # remain on 'main' branch if target branch is 'master'
```
[the remark this answers is not in the document]
Comment (feedmeapples) [reply .github/workflows/update-submodules.yml]: +

**Review by dnr — APPROVED** [review]

**Pull request merged**